File: src/ExampleTableFormatter.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class ExampleTableFormatter {

    private final String DEL = "|";
    private final String DEL_REGEXP = "\\|";
    private int currentTableCellsCount = 0;

    public List<String> formatExampleTablesOfSourceLines(final LinkedList<String> sourceLines) throws IOException {
        boolean tableLineCollecting = false;
        int startIndex = 0;
        for (int lineIndex = 0; lineIndex < sourceLines.size(); lineIndex++) {
            String l = sourceLines.get(lineIndex);
            if (tableLineCollecting && !isCurrentTableLine(l)) {
                formatTable(sourceLines.subList(startIndex, lineIndex));
                tableLineCollecting = false;
            }
            if (!tableLineCollecting && isTableLine(l)) {
                tableLineCollecting = true;
                currentTableCellsCount = l.split(DEL_REGEXP).length;
                startIndex = lineIndex;
            }
            if (tableLineCollecting && lineIndex == sourceLines.size() - 1) {
                formatTable(sourceLines.subList(startIndex, sourceLines.size()));
                tableLineCollecting = false;
            }
        }
        return sourceLines;
    }

    private void formatTable(List<String> exampleTable) {
        Map<Integer, Integer> columnLengthMap = getColumnLengthMap(exampleTable);
        for (int lineIndex = 0; lineIndex < exampleTable.size(); lineIndex++) {
            List<Cell> cells = getCellsFromTableLine(exampleTable.get(lineIndex));
            exampleTable.remove(lineIndex);
            exampleTable.add(lineIndex, DEL + cells.stream().map(
                    c -> c.getContent() + generateSpaces(columnLengthMap.get(cells.indexOf(c)) - c.getContentLength()))
                    .collect(Collectors.joining(DEL)) + DEL);
        }
    }

    private Map<Integer, Integer> getColumnLengthMap(List<String> exampleTable) {
        Map<Integer, Integer> columnLength = new HashMap<>();
        exampleTable.forEach(line -> {
            List<Cell> cells = getCellsFromTableLine(line);
            for (int cellIndex = 0; cellIndex < cells.size(); cellIndex++) {
                Cell cell = cells.get(cellIndex);
                columnLength.merge(cellIndex, cell.getContentLength(), (preMaxLength,
                        maxLengthCandidate) -> maxLengthCandidate > preMaxLength ? maxLengthCandidate : preMaxLength);
            }
        });
        return columnLength;
    }

    private boolean isTableLine(final String str) {
        return isStringMatchesRegExp("^" + DEL_REGEXP + ".*" + DEL_REGEXP, str);
    }

    private boolean isCurrentTableLine(final String str) {
        return isTableLine(str) && str.split(DEL_REGEXP).length == currentTableCellsCount;
    }

    private List<Cell> getCellsFromTableLine(final String line) {
        String[] cellsArr = line.split(DEL_REGEXP);
        List<Cell> cells = new ArrayList<>();
        for (int cellIndex = 0; cellIndex < cellsArr.length; cellIndex++) {
            cells.add(new Cell(cellsArr[cellIndex].trim()));
        }
        if (!cells.isEmpty()) {
            cells.remove(0); // Removing first empty cell appearing after split
        }
        return cells;
    }

    private String generateSpaces(final int spaceCount) {
        return new String(new char[spaceCount]).replace("\0", " ");
    }

    public boolean isStringMatchesRegExp(final String regExp, final String str) {
        Pattern pattern = Pattern.compile(regExp);
        Matcher matcher = pattern.matcher(str);
        return matcher.find();
    }
}

File: src/main/java/ExampleTableFormatter.kt

import java.util.*
import java.util.regex.Pattern


private val DEL = "|"
private val DEL_REGEXP = """\|""".toRegex()


fun jBehaveTablesOfSourceLines(sourceLines: LinkedList<String>): List<String> {
    var tableLineCollecting = false
    var startIndex = 0
    for (lineIndex in sourceLines.indices) {
        val l = sourceLines[lineIndex]
        if (tableLineCollecting && !isTableLine(l)) {
            formatTable(sourceLines.subList(startIndex, lineIndex))
            tableLineCollecting = false
        }
        if (!tableLineCollecting && isTableLine(l)) {
            tableLineCollecting = true
            startIndex = lineIndex
        }
        if (tableLineCollecting && lineIndex == sourceLines.size - 1) {
            formatTable(sourceLines.subList(startIndex, sourceLines.size))
            tableLineCollecting = false
        }
    }
    return sourceLines
}

private fun getTables(sourceLines: LinkedList<String>): List<MutableList<String>>{
    val tables = mutableListOf<MutableList<String>>()

    var tableLineCollecting = false
    var startIndex = 0
    for (lineIndex in sourceLines.indices) {
        val l = sourceLines[lineIndex]
        if (tableLineCollecting && !isTableLine(l)) {
            tables.add(sourceLines.subList(startIndex, lineIndex))
            tableLineCollecting = false
        }
        if (!tableLineCollecting && isTableLine(l)) {
            tableLineCollecting = true
            startIndex = lineIndex
        }
        if (tableLineCollecting && lineIndex == sourceLines.size - 1) {
            tables.add(sourceLines.subList(startIndex, sourceLines.size))
            tableLineCollecting = false
        }
    }
    return tables
}

private fun formatTable(table: MutableList<String>) {
    val offset = table.firstOrNull()?.takeWhile(Char::isWhitespace)?.length ?: 0
    val columnLengthMap = getColumnLengthMap(table)
    table.map { line -> formatTableLine(line, columnLengthMap, offset) }.toList(table)
}

private fun formatTableLine(line: String, columnLengthMap: Map<Int, Int>, offset: Int) : String{
    val cells = getCellsFromTableLine(line)
    val prefix = " ".repeat(offset) + DEL + " "
    return columnLengthMap.map { (index, length) ->
        val cell = cells.getOrNull(index) ?: Cell("")
        cell.content.padEnd(length)
    }.joinToString(" $DEL ", prefix, " $DEL")
}

private fun getColumnLengthMap(table: List<String>): Map<Int, Int> {
    val columnLengthMap = mutableMapOf<Int, Int>()
    table
        .map { tableLine -> getCellsFromTableLine(tableLine) }
            .forEach{ lineCells ->
                lineCells.forEachIndexed{ index, cell ->
                    if(cell.contentLength > columnLengthMap[index] ?: 0){
                        columnLengthMap[index] = cell.contentLength
                    }
                }
            }
    return columnLengthMap
}

private fun isTableLine(str: String): Boolean {
    return isStringMatchesRegExp("""^\s*$DEL_REGEXP.*$DEL_REGEXP""", str)
}

private fun getCellsFromTableLine(line: String): List<Cell> {
    return line.split(DEL_REGEXP)
            .map { it.trim{ it <= ' '} }
            .dropWhile { it.isEmpty() }
            .dropLastWhile { it.isEmpty() }
            .map { Cell(it) }
}

fun isStringMatchesRegExp(regExp: String, str: String): Boolean {
    val pattern = Pattern.compile(regExp)
    val matcher = pattern.matcher(str)
    return matcher.find()
}

/**
 * overwrite list by this list content
 */
fun <T> List<T>.toList(list: MutableList<T>){
    list.clear()
    this.forEach { list.add(it) }
}

